### ISSUE TITLE
Forge configs occasionally become corrupted

### ISSUE THREAD

**Issue by embeddedt** [neoforged/NeoForge#32]
This is a continuation of https://github.com/MinecraftForge/MinecraftForge/issues/9122.

After performing an analysis of the code Forge uses to interact with the Night Config library, I identified a design flaw which seems to have the potential for race conditions and corruption. In `ConfigFileTypeHandler` when a config is first registered and loaded, a file watcher is [attached](https://github.com/MinecraftForge/MinecraftForge/blob/30f255d76c4c2e915414a24e6d8aed42f429817e/fmlcore/src/main/java/net/minecraftforge/fml/config/ConfigFileTypeHandler.java#L51). Shortly after that method returns [`save` is invoked on the config object](https://github.com/MinecraftForge/MinecraftForge/blob/30f255d76c4c2e915414a24e6d8aed42f429817e/fmlcore/src/main/java/net/minecraftforge/fml/config/ConfigTracker.java#L63C15-L63C15). In summary, this can lead to the following sequence of events:

1. The CommentedFileConfig instance is built and the watcher is registered (the file on disk is now being watched for changes)
2. Forge code invokes config.save() which calls into Night Config to write to the file synchronously.
3. As part of the process of saving, `Files.newOutputStream` is invoked with the config file's path, which, according to the [Javadoc](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/nio/file/Files.html#newOutputStream(java.nio.file.Path,java.nio.file.OpenOption...)), "initially [truncates] an existing regular-file to a size of 0 if it exists." I verified using a simple test program that, at least on Linux, the `OutputStream` given by this method is *not* atomic. The file is truncated immediately upon return of the method, and writes are seen in the underlying file even before the stream is closed.
4. Because Night Config's file watcher is asynchronous, it will detect any changes to the underlying file, regardless of whether or not saving has actually completed. That means it may see the fully truncated file, or a half-written file. In this scenario, attempting to read from the incomplete file is very likely to result in random errors when the system encounters some corrupted version.

When testing on 1.16, I was able to reliably observe symptoms similar to the theory I have described above - by augmenting the Forge file watcher's `run` function with my own, I confirmed that the [`currentlyWriting`](https://github.com/TheElectronWill/night-config/blob/e8c9f74808fe412dc43464370b206a1fe893d493/core/src/main/java/com/electronwill/nightconfig/core/file/WriteSyncFileConfig.java#L29) flag in `WriteSyncFileConfig` (the config class in NC responsible for saving) was still `true` when the watcher was invoked, indicating that saving is NOT completed before the watcher is fired.

I think fully fixing this will require a fundamental rethink of how the file watcher is used in Forge, as ideally it needs to never be listening when `.save` is called.

For completeness, [here is a link to the Discord conversation](https://discord.com/channels/313125603924639766/852298000042164244/1130678626933813340) in which I first brought up this new theory.

**Comment by quat1024** [comment]
It doesn't help that nightconfig's filewatcher is apparently really aggressive and checks for file changes up to [a million times a second apparently](https://github.com/TheElectronWill/night-config/pull/144) (???) which could exacerbate this race

**Comment by TheElectronWill** [comment]
Hi! NightConfig's author here.
First of all, thanks for reporting the issue. I cannot access the discord's conversation, do you have an invitation to the relevant server?
The `currentlyWriting` flag is meant to avoid that: if `save()` hasn't finished yet, `load()` will do nothing. It's not a problem if the FileWatcher calls the handler, because `load()` won't do anything... at least that's the theory.

~~In practice, I suspect that the problem comes from the `load()` method: the double-checked locking idiom is badly implemented:~~
```java
	@Override
	public void load() {
		if (!currentlyWriting) {
			synchronized (this) {
                                // HERE is missing a double-check if(!currentlyWriting)
				if (closed) {
					throw new IllegalStateException("Cannot (re)load a closed FileConfig");
				}
				parser.parse(nioPath, config, parsingMode, nefAction);
			}
		}
	}
```
edit: but since write() and parse() are in a synchronized block, reading and writing still isn't possible at the same time. So it's weird :thinking: 

As for the FileWatcher, that shouldn't cause the problem. A partial write of the file and/or partially updated config may be the root cause instead.

**Comment by AterAnimAvis** [comment]
It's the NeoForge (previously Forge) Discord Server, https://discord.neoforged.net

**Comment by TheElectronWill** [comment]
thanks!

**Comment by TheElectronWill** [comment]
Update: https://github.com/TheElectronWill/night-config/pull/152 has been merged with a full rewrite of FileConfig, based on concurrent configurations (see the comments on the PR). When the new API is used properly, it should hopefully fix the problem :)

**Comment by TheElectronWill** [comment]
[NightConfig v3.7.0](https://github.com/TheElectronWill/night-config/releases/tag/v3.7.0) released :tada:

**Comment by ljfa-ag** [comment]
Regarding the difficulty of reproducing this issue, maybe this could be helpful:

I've noticed that corruption happens far more often when the config values are manually loaded early by reflecting into `ConfigTracker.openConfig()` (I know this is discouraged, but it is a reality that some mods are doing this. Fortunately, https://github.com/neoforged/FancyModLoader/pull/150 is a thing  now.) After trying 4 or 5 times, the config that is early loaded this way becomes reset to default on my Linux machine with Forge 43.3.0 on Minecraft 1.19.2 (haven't tested it with NeoForge yet). On subsequent runs, the resets happen pretty reliably.

Calling `ConfigTracker.openConfig()` early this way adds a ConfigWatcher before the config is loaded at the proper time later, the presence of which seems to exacerbate the issue. When I manually remove the watcher by calling `FileWatcher.defaultInstance().removeWatch()`, the resets seem to disappear or at least happen far less frequently. It should be noted that adding a watcher to a file that is already being watched will not replace the original watch handler, but I don't know if this is of relevance here.

**Comment by TheElectronWill** [comment]
This seems to confirm that the root cause of the issue is a bad interaction between autoloading, loading and using the config. 
 
When NeoForge will upgrade its dependency, it should go away.

**Comment by ljfa-ag** [comment]
As far as I can tell, the NightConfig update should also fix another race condition that I found that is not directly related to file I/O:

While the WriteSyncFileConfig is being loaded, it is still possible to access the config values. In this case, you might get an empty or partially parsed config. The following could happen:
1. The CommentedFileConfig instance is built, the config is loaded from the disk and the watcher is registered.
2. Forge code invokes config.save() which calls into Night Config to write to the file synchronously. _Let's suppose that the write happens atomically and no empty or partially written file can be observed._
3. The file watcher thread detects the file being written and causes the config to reload, calling `TomlParser.parse()`. The parsing will initially clear the config with `ParsingMode.REPLACE`.
4. If the config value is accessed by the main thread via `config.getRaw()` at this moment, it will return null since the config was cleared.
5. The config is re-parsed and restored to the proper state by the file watcher thread. Calling `config.getRaw()` now will return the correct value again.

Note that no file corruption is necessarily going on here, but I've still had this manifest occasionally as default values being returned when accessing the config right after early loading it. Of course this is not limited to early loading but can also happen when e.g. listening to a ModConfigEvent.

Looking at the changes in https://github.com/TheElectronWill/night-config/pull/152, it seems like this shouldn't be possible anymore, since `SyncFileConfig.load()` will replace the config atomically and the access methods in `SynchronizedConfig.getRaw()` are also synchronized on the same monitor. I still wanted to post this here since I spent way too much time going down the rabbit hole and debugging this issue :)
